File: python/algo/202508/algo0222.py
from bisect import bisect_left
from typing import Optional
from python.common import TreeNode

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# 二叉树 - 2.16 其他
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        # 从 c++ 转换过来，使用 bisect_left 库函数
        h = 0
        if root is None: return 0
        cur = root
        while cur.left is not None:
            h += 1
            cur = cur.left
        
        left, right = (1 << h) - 1, 1 << (h+1)
        def check(mid: int) -> bool:
            cur = root
            for i in range(h-1, -1, -1):
                if mid >> i & 1:
                    if cur.right is not None: cur = cur.right
                    else: return False 
                else:
                    if cur.left is not None: cur = cur.left
                    else: return False 
            return True
        left, right = (1 << h) - 1, 1 << (h+1)
        while left + 1 < right:
            mid = (left + right) >> 1
            if check(mid):
                left = mid
            else: right = mid
        return left
        # 下面库函数不正确 ???
        # return bisect_left(range(left, right), True, key=check) + left

File: cplus/leetcode/202508/algo0222.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// 二叉树 - 2.16 其他
class Solution {
public:
    int countNodes(TreeNode* root) {
        // 参考题解，二分查找
        int h = 0;
        if (!root) return 0;
        auto cur = root;
        while (cur->left) {h++, cur= cur->left;}

        auto left = (1 << h) - 1, right = 1 << (h+1);
        
        auto check = [&](int m) -> bool {
            auto cur = root;
            for (int i=h; i; --i) {
                if (m >> (i-1) & 1) {
                    if (cur->right) cur = cur->right;
                    else return false;
                } else {
                    if (cur->left) cur = cur->left;
                    else return false;
                }
            }
            
            return true;
        };

        while (left + 1 < right) { // 区间不为空
            int mid = left + (right - left) / 2;
            if (check(mid)) {
                left = mid; // 范围缩小到 (mid, right)
            } else {
                right = mid; // 范围缩小到 (left, mid)
            }
        }
        return left;

    }
};